admin_apps/shared_utils.py

```diff
@@ -823,6 +828,363 @@ def download_yaml(file_name: str, conn: SnowflakeConnection) -> str:
             # Read the raw contents from {temp_dir}/{file_name} and return it as a string.
             yaml_str = temp_file.read()
             return yaml_str
+        
+def unpack_yaml(data):
+    """
+    Recursively unpacks a YAML structure into a Python dictionary.
+    """
+    if isinstance(data, dict):
+        return {key: unpack_yaml(value) for key, value in data.items()}
+    elif isinstance(data, list):
+        return [unpack_yaml(item) for item in data]
+    else:
+        return data
+
+def load_yaml_file(file_paths) -> list[dict]:
+    """
+    Loads one or more YAML files and combines them into a single list.
+    """
+    combined_yaml = []
+    for file_path in file_paths:
+        yaml_content = yaml.safe_load(file_path)
+        combined_yaml.append(unpack_yaml(yaml_content))
+    return combined_yaml
+
+def extract_key_values(data: list[dict], key: str) -> list[dict]:
+    """
+    Extracts key's value from a list of dictionaries.
+    """
+    result = []
+    for item in data:
+        values = item.get(key, [])
+        if isinstance(values, list):
+            result.extend(values)
+        else:
+            result.append(values)
+    return result
+
+def extract_expressions_from_sections(data_dict, section_names):
+    """
+    Extracts data in section_names from a dictionary into a nested dictionary:
+    """
+    def extract_key(obj):
+        return obj.get('expr', obj['name']).lower()
+    
+    d = {}
+    for i in section_names:
+        d[i] = {extract_key(obj): obj for obj in data_dict.get(i, [])}
+    
+    return d
+
+def upload_label():
+    if st.session_state.get('my_file_uploader', None):
+        st.session_state['upload_label'] = ', '.join([i.name for i in st.session_state['my_file_uploader']])
+    elif st.session_state.get('uploaded_semantic_files', None):
+        st.session_state['upload_label'] = ', '.join([i for i in st.session_state['uploaded_semantic_files']])
+    else:
+        st.session_state['upload_label'] = f'Upload {st.session_state["partner_tool"]} semantic yaml file(s)'
+
+
+def upload_partner_semantic() -> bool:
+    """
+    Upload the semantic model to a stage.
```

Comment:
to clarify - this doesn't actually upload the file to a snowflake stage, right? just imports the contents locally?